site/src/contexts/ProxyContext.tsx

```diff
@@ -66,41 +97,39 @@ export const ProxyProvider: FC<PropsWithChildren> = ({ children }) => {
   } = useQuery({
     queryKey,
     queryFn: getWorkspaceProxies,
-    // This onSuccess ensures the local storage is synchronized with the
-    // proxies returned by coderd. If the selected proxy is not in the list,
-    // then the user selection is removed.
-    onSuccess: (resp) => {
-      setAndSaveProxy(proxy.selectedProxy, resp.regions)
-    },
   })
 
-  // Everytime we get a new proxiesResponse, update the latency check
+  // Every time we get a new proxiesResponse, update the latency check
   // to each workspace proxy.
   const proxyLatencies = useProxyLatency(proxiesResp)
 
-  const setAndSaveProxy = (
-    selectedProxy?: Region,
-    // By default the proxies come from the api call above.
-    // Allow the caller to override this if they have a more up
-    // to date list of proxies.
-    proxies: Region[] = proxiesResp?.regions || [],
-  ) => {
-    if (!proxies) {
-      throw new Error(
-        "proxies are not yet loaded, so selecting a proxy makes no sense. How did you get here?",
-      )
-    }
-    const preferred = getPreferredProxy(proxies, selectedProxy)
-    // Save to local storage to persist the user's preference across reloads
-    // and other tabs.
-    savePreferredProxy(preferred)
-    // Set the state for the current context.
-    setProxy(preferred)
-  }
+  // updateProxy is a helper function that when called will
+  // update the proxy being used.
+  const updateProxy = useCallback(() => {
+    // Update the saved user proxy for the caller.
+    setUserSavedProxy(loadUserSelectedProxy())
+    setProxy(
+      getPreferredProxy(
+        proxiesResp?.regions ?? [],
+        // For some reason if I use 'userSavedProxy' here,
+        // the tests fail. It does not load "undefined" after a
+        // clear, but takes the previous value.
+        loadUserSelectedProxy(),
```

Comment:
`setState`s are async - when you set a value, it isn't set instantly. So `userSavedProxy` may not be updated by the time you try to grab it here (also, I don't really see where we're updating it - seems like we're just setting it to the value of what is already in local storage). That is probably why you're seeing the previous value.

Reply:
I tried adding an `await setUserSaveProxy` and still had this behavior. 

Ln 110 only loads from local storage, it does not do any writes.

Reply:
My bad about ln 110 - was confusing `setUserSavedProxy` and `saveUserSelectedProxy` as mentioned above.

Regarding your `await`: you can't await a `setState` because, although asynchronous, it doesn't return a promise. You could write a callback but I think what you have here is best. Do you mind removing the comment? I think this is expected behavior.